sdk/trace/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java

```diff
@@ -178,44 +191,58 @@ private Worker(
     }
 
     private void addSpan(ReadableSpan span) {
-      if (!queue.offer(span)) {
+      if (queueSize.get() >= maxQueueSize) {
         droppedSpans.add(1);
+      } else {
+        queue.offer(span);
+        queueSize.incrementAndGet();
+        if (addedSpansCounter.incrementAndGet() % maxExportBatchSize == 0) {
+          lock.lock();
+          try {
+            needExport.signal();
+          } finally {
+            lock.unlock();
+          }
+        }
       }
     }
 
     @Override
     public void run() {
       updateNextExportTime();
-
       while (continueWork) {
         if (flushRequested.get() != null) {
           flush();
         }
-
-        try {
-          ReadableSpan lastElement = queue.poll(100, TimeUnit.MILLISECONDS);
-          if (lastElement != null) {
-            batch.add(lastElement.toSpanData());
-          }
-        } catch (InterruptedException e) {
-          Thread.currentThread().interrupt();
-          return;
+        while (!queue.isEmpty() && batch.size() < maxExportBatchSize) {
+          queueSize.decrementAndGet();
+          batch.add(queue.poll().toSpanData());
         }
-
         if (batch.size() >= maxExportBatchSize || System.nanoTime() >= nextExportTime) {
           exportCurrentBatch();
           updateNextExportTime();
         }
+        if (queue.isEmpty()) {
+          lock.lock();
+          try {
+            long pollWaitTime = nextExportTime - System.nanoTime();
+            if (pollWaitTime > 0) {
+              needExport.awaitNanos(pollWaitTime);
+            }
+          } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            return;
+          } finally {
+            lock.unlock();
+          }
+        }
       }
     }
 
     private void flush() {
-      int spansToFlush = queue.size();
-      while (spansToFlush > 0) {
-        ReadableSpan span = queue.poll();
-        assert span != null;
-        batch.add(span.toSpanData());
-        spansToFlush--;
+      while (!queue.isEmpty()) {
+        batch.add(queue.poll().toSpanData());
+        queueSize.decrementAndGet();
```

Comment:
Is there a reason why ithe polling and the decrementing the queuesize are done in the opposite order in this method than in the main run loop?

Reply:
The order doesn't matter here since queueSize is atomic and will be consistent

Reply:
then please put them in the same order, or extract a method that does both operations so it's not confusing